crates/proto/src/rr/record_data.rs

```diff
@@ -37,6 +37,20 @@ pub trait RecordData: Clone + Sized + PartialEq + Eq + fmt::Display {
     /// Attempts to convert to this RecordData from the RData type, if it is not the correct type the original is returned
     fn try_from_rdata(data: RData) -> Result<Self, RData>;
 
+    /// Read the RecordData from the data stream.
+    ///
+    /// * `decoder` - data stream from which the RData will be read
+    /// * `record_type` - specifies the RecordType that has already been read from the stream
+    /// * `length` - the data length that should be read from the stream for this RecordData
+    fn read(
+        decoder: &mut BinDecoder<'_>,
+        record_type: RecordType,
```

Comment:
During deserialization the RecordType isn't known, and since RData itself implements RecordData, we need to know the type. I'm not sure that there's a simple way to separate these things? I'll double check.

Reply:
Hmm, yeah, why does `RData` implement `RecordData`? Seems like possibly an odd design choice.

Reply:
So the specific reason is that `RData` is a field in `Record` and we restrict `Record<T: RecordData>`, so `RData` then needs to implement `RecordData`... There's a bit of a rabbit hole here where that `RecordData` then requires `BinDecodable`, and thus needing the `RecordType` during decoding... I'm trying to think of how to clean that up easily, but it's not currently obvious.

Reply:
Ok, I cleaned this up. `Record` can only be decoded with `RData` now, but that's the only implementation that makes sense anyway. I think this closes out the final issue in this review.